### ISSUE TITLE
Collision Detection Can Fail When Instances Move

### ISSUE THREAD

**Issue by mistletoe** [YoYoGames/GameMaker-Bugs#4743]
### Description

OK, as per YYG's request for the last report... here's a new short bug report and an attached project demonstrating how it happens. 

Basically, collision_line() can fail under some very odd circumstances having to do with Instances being moved. 

It's not collision_line() itself, so far as I can determine. 

It's something else that happens when Instances are moved in both X and Y during the same frame.

### Expected Change

Please investigate and fix.

### Steps To Reproduce

1. Start GameMaker
2. Build project attached.
3. Run project.
4. See the issue
5. ...profit?
[COLLISION FAIL TEST.zip](https://github.com/YoYoGames/GameMaker-Bugs/files/14300844/COLLISION.FAIL.TEST.zip)

### How reliably can you recreate this issue using your steps above?

Always

### Which version of GameMaker are you reporting this issue for?

2023.11.1 (Monthly)

### Which platform(s) are you seeing the problem on?

Windows

### Contact Us Package Attached?

- [X] I have attached my Contact Us Package

### Sample Project Added?

- [X] I have included a small sample project

**Comment by mistletoe** [comment]
Note:

If you want to try this project out and see what I'm seeing (what should happen, vs. what does) you can uncomment the draw code and comment out the bit that only draws the failure states.

Basically, this project does 30K collision_line() operations a frame (dial that back, if you're running this on slower hardware than mine, until you're somewhere around 40 FPS).

It's firing the test against a target that's supposed to wiggle, turn and rescale in the middle of the screen.

Where things got _weird_: everything worked as advertised, until I started wiggling the target around more than 3 pixels in XY. Then failure conditions were occurring, but it wasn't until I set up the draw code that it was obvious that something very strange occurs. 

Why 3 pixels? IDK, but probably this is when near-parallel conditions cause a near-miss. 

Why is X at the correct position, but Y clearly isn't, in the failure states? IDK. The code looks right...

Whatever this is... it's rare. Like, really, really rare; you'll have to run this for a few seconds to see failure states (and if you're on a small resolution, IDK whether it'll be obvious or not; I set up the Room at 2048X2048). I'm only seeing it because I have a project with a fairly-large number of collision_line() checks from things that are in motion every frame.

**Comment by mistletoe** [comment]
Here's a second build that'll be helpful for people trying to understand the error. 

With this version, I've done a manual correction of the Instance's position, for display purposes. You can see that there are times when it's failing the test, even if XY were where they should have been. So clearly XY aren't where they're supposed to be; it's not just an illusion.

Figuring out how to correct the issue so that the display is "where things should have happened" was a little puzzle, because you simply cannot just plug in the change to Y; it's consistently wrong. But it confirms that the first project's code operates correctly and produces these error states.

[COLLISION FAIL TEST 2.zip](https://github.com/YoYoGames/GameMaker-Bugs/files/14301613/COLLISION.FAIL.TEST.2.zip)

**Comment by attic-stuff** [comment]
tldr; i cannot reproduce your error.

you do not have precise collisions set to true for the collision_line call. what this means is that collision_line will check the bounding box (which never rotates) instead of the mask and this is going to give false positive results when your expectation is essentially a thick line intersecting a thin line. if your mask is going to be rotated and you want to retain the precise nature of that mask (because its a precise mask or rectangle with rotation mask) then you need to make sure the precise argument is true in the collision_line call. from the [manual](https://manual.gamemaker.io/monthly/en/GameMaker_Language/GML_Reference/Movement_And_Collisions/Collisions/collision_line.htm): 

> for precise collisions to be considered both the object sprite and the collision function must have precise marked as on.

once you set that to true you will notice that your collision line hits even LESS often, because of how precise the check is. this is kind of hard to visualize in your project because you're trying to draw a 1 pixel wide line at a fragment resolution of 2048x2048.

i have put together a much much smaller project for you to play with. basically what this project does, is transform and translate a 2 pixel wide rectangle every 40 frames, while testing for a collision with that rectangle against a collision line that runs from corner to corner of the room. you'll notice how extremely close the rectangle can get to the line without it triggering a collision, which turns the collision line red. it also has a snapshot ability, which will run the project for 2 minutes, transform the rectangle every 2 frames and snapshot every failed collision. if you review those snapshots (%localappdata%\collision_line_test\) then you will see just how close the line can get to the rectangle without failing. here is a gif i made of all kinds of tests it ran where the lines snuggle REAL close but do not collide:
![linetest](https://github.com/YoYoGames/GameMaker-Bugs/assets/44242527/3b923273-9d7e-4113-900a-181b17c553a4)

here is the project for playing with: [collision line test.zip](https://github.com/YoYoGames/GameMaker-Bugs/files/14302535/collision.line.test.zip)

p.s. gm is locked into step, which means it will never skip calculations or develop imprecision because the fps drops due to there being a godzillion instances or calculations. in your included project you will get the same results with 1 check per step instead of 30,000 and you will get the same results with no BLANKs in the level.

**Comment by mistletoe** [comment]
OK, this weird. I set it to true, and it's still producing the errors here.

**Comment by mistletoe** [comment]
...and here's a version where it shows the offset still happening, lol.  
[COLLISION FAIL TEST 4.zip](https://github.com/YoYoGames/GameMaker-Bugs/files/14302862/COLLISION.FAIL.TEST.4.zip)

I'll look at your code, but I'm intrigued that you're saying you didn't see this:

![collision_error](https://github.com/YoYoGames/GameMaker-Bugs/assets/647639/1fdd63fb-3cc8-42b9-8d77-fbe9aba3b9b4)

**Comment by mistletoe** [comment]
OK. Looked at your code, and with minor changes, it's reproducing my results. This code changes your logic so that when there's a failure (i.e., a result of noone) then the line and the rectangle should be drawn. Note that the rectangle doesn't draw, because once again, it's not where it's supposed to be. The black rectangle isn't drawn, because it's somewhere else.

[collision line test 2.zip](https://github.com/YoYoGames/GameMaker-Bugs/files/14303048/collision.line.test.2.zip)

When the condition in Draw is reversed, I get this, which is what's expected:

![collision_error2](https://github.com/YoYoGames/GameMaker-Bugs/assets/647639/b693fa27-9104-4624-be8c-ceff1336a9c2)

**Comment by mistletoe** [comment]
Now here's a version of your code where I'm storing all of the data into an array, like my first example, and moving it in Draw. Now we can see the sprite draw, because it's where it should be. This is where things got intriguing. Most of these results look OK to me... _after_ I do that crucial step of offsetting. 

Without the offset stage, you'd not see this result at all.
[collision line test 3.zip](https://github.com/YoYoGames/GameMaker-Bugs/files/14303313/collision.line.test.3.zip)

**Comment by mistletoe** [comment]
So... erm... 

1. The actual collision_line() results look fine. What appears to be happening is that objects are getting offset when moved, or their sprite masks are still somewhere else.
2. There's still something weird going on when we move things.
3. When I correct for that, so that we can see where the Instance should have been, rather than where GMS thought it was, in this smaller Room size, I see results that (except for the offset) appear to be arguably fine. 
4. But in the first system above, where I put everything under a bunch of load and a lot more resolution, I see more failures, and worse ones. Despite having slightly different implementations, these are all just doing the same things.
5. The simpler system isn't producing _zero_ errors, but I'm not seeing obvious trace-through like occasionally in the example #2. Arguably, some misses at edges. 

That's actually _less_ helpful, because I can tell you right now, I see trace-through failures often enough to have filed this bug report in the first place. I'm not crazy, folks.

In the first two examples, it appears that the Y is 1024 higher than it should be. Or maybe it's set to zero briefly because we touched the X in Step? I don't want to have to do two raycasts, one at the place it should be, one where I presume Y is 0, though; that may cause other problems and it's hugely redundant.

**Comment by attic-stuff** [comment]
you have a typo causing this error in your project. you wrote barVars instead of badVars

**Comment by mistletoe** [comment]
Uh.... line 55 is fine:

<img width="364" alt="image" src="https://github.com/YoYoGames/GameMaker-Bugs/assets/647639/5c111f19-bc5c-4cb3-b7f3-e54548b4e25d">

**Comment by mistletoe** [comment]
OK. Looked at that a bit more, fixed that variable, etc. That makes the results look like the version of yours I've modified, agreed. So we're still dealing with:

1. There may still be something weird happening with the results when moved. Most of the results I'm seeing in that last one of yours I modified looks legit... once it's moved.
2. Whatever it is, we're not seeing it if we record the proper offset, but are seeing it if we go about it via draw calls.
3. collision_line() looks OK. Whatever's going on isn't there.
4. I'm still seeing whatever-this-is happen in my actual production code. I have enemies that occasionally see through walls, guided missiles that occasionally target enemies they shouldn't be able to see, it's something I've had to work around for other areas of the game. I never trust the results at this point. Maybe I should just get over it and write a collision system in GML that handles rotated rectangles at a reasonable speed, since that's the major issue I see over and over.

**Comment by mistletoe** [comment]
OK, here's a new version of your code, but it's using a method to detect collisions with rotated rectangles, instead of Precise collisions. It uses [xot's line-intersection code](https://www.gmlscripts.com/script/lines_intersect) for one part, and some other stuff I wrote.

It's faster than I thought it would be (0.02 ms when profiled here, roughly on-par with pixel-perfect collision_line())... and it looks like it works properly. 

Honestly, though, whatever this problem is... I don't even think it's collision_line() now. But _something's_ going on with collision detection, and it's happening often enough that it's noticeable in testing.

I'm going to review that guided-missile logic that started this... and see if there's anything else to check out. This issue keeps cropping up in the strangest places.

[collision line test 4.zip](https://github.com/YoYoGames/GameMaker-Bugs/files/14307477/collision.line.test.4.zip)

**Comment by mistletoe** [comment]
I've provided one last version of this; after number 4, it seemed like I should just build a proper raycaster and sorter. 

Now it's performing all of the functionality of scr_rangefinder_exact() with anything that's a rectangle. Rotated, scaled, whatever.  

This is:

1. A raycast solution; it's using line-section intersections. Draw events document what has happened. A mouse-click resets the simulation and allows you to see another solution.
2. A distance sorter that will provide the closest hit location and Instance id.
3. It takes about 10 microseconds to compute per Instance; a test with 51 Instances averages somewhere around 500 microseconds. Scaling is linear, but inherently unpredictable; my results indicate about 10% time-variance.
4. I'm licensing this MIT; do whatever you want.
5. As a bonus, by setting "exitEarly" to true, you can get this to behave much like collision_line() with any rotated rectangle, where the first hit recorded will stop iteration and return the hit location and the id of the Instance it intersected. This can reduce time with 51 objects considerably (~80 microseconds).

[collision line test 5.zip](https://github.com/YoYoGames/GameMaker-Bugs/files/14317757/collision.line.test.5.zip)

License for the scripts is MIT.

**Comment by mistletoe** [comment]
OK, not done after all. Had an idea this morning for a simple optimization; uses a square-test to eliminate most possible targets. Will not speed this up for situations where all targets are within the rectangle, but for more-typical situations, it speeds things up a lot: ~80 microseconds or so. I've also taken the time to make the diagnostics in the project easier to interpret, etc., and the code now returns the id, XY and distance, in that order, like we're used to.

I'm probably going to extend this to support rotated polygons. As things stand, I suspect this is about to speed my game's performance up by a lot for certain tasks I'm performing on a regular basis, and in initial testing, it appears to be at least as reliable. This is considerably faster than the pixel-perfect alternative I've been using for years. 

Code license is MIT.

[collision line test 6.zip](https://github.com/YoYoGames/GameMaker-Bugs/files/14320219/collision.line.test.6.zip)

**Comment by attic-stuff** [comment]
is there a bug you need to report here with all of these projects still? remember the topic of this issue was a bug in collision line, which was not a bug in collision_line but in your code and if you no longer have this bug you should close the issue proper. if you do come across the actual bug you think may be happening, it needs to be reported with a project so it can be reproduced.

**Comment by mistletoe** [comment]
I'll close it. So far as I can tell, collision_line() works. Whatever's going on remains a mystery, but it's not here. I want all these raycasts in GMS, though; it'd probably run considerably faster in C#.
